Tools/clinic/clinic.py

```diff
@@ -2270,44 +2275,61 @@ def compute_checksum(input, length=None):
 
 
 class PythonParser:
-    def __init__(self, clinic):
+    def __init__(self, clinic: Clinic) -> None:
         pass
 
-    def parse(self, block):
+    def parse(self, block: Block) -> None:
         s = io.StringIO()
         with OverrideStdioWith(s):
             exec(block.input)
         block.output = s.getvalue()
 
 
 class Module:
-    def __init__(self, name, module=None):
+    def __init__(
+            self,
+            name: str,
+            module: Module | None = None
+    ) -> None:
         self.name = name
         self.module = self.parent = module
 
-        self.modules = collections.OrderedDict()
-        self.classes = collections.OrderedDict()
-        self.functions = []
+        self.modules: ModuleDict = collections.OrderedDict()
+        self.classes: ClassDict = collections.OrderedDict()
+        self.functions: list[Function] = []
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return "<clinic.Module " + repr(self.name) + " at " + str(id(self)) + ">"
 
+ModuleDict = dict[str, Module]
```

Comment:
Nit: For readability, I'd prefer it if this alias could be defined above the `Module` class, since it's used inside the `Module` class for annotations. You can use strings to get around the fact that `Module` hasn't yet been defined:

```py
ModuleDict = dict[str, "Module"]
```

(Same point for `ClassDict` below; I'd prefer it if it were defined before it's used as an annotation.)

Reply:
Ah, neat. TIL.